File: examples/keyboardControl.cpp
#include "airbase.hpp"
#include "argparse/argparse.hpp"

int main(int argc, const char *argv[]) {
  argparse::ArgumentParser program("airbase");
  program.add_argument("-mp", "--map_path").help("the .stcm map path to build/load").default_value("map.stcm");
  program.add_argument("-sp", "--speed")
      .help("the speed of the RobotBase")
      .default_value("high")
      .choices("low", "medium", "high");
  program.add_argument("-ip", "--ip_address").help("the ip address of RobotBase").default_value("192.168.11.1");
  try {
    program.parse_args(argc, const_cast<char **>(argv));
  } catch (const std::runtime_error &err) {
    std::cout << err.what() << std::endl;
    std::cout << program.help().str();
    exit(0);
  }

  std::string map_path = program.get("-mp");
  std::string speed = program.get("-sp");
  std::string ip = program.get("-ip");

  AirBase airbase(ip);

  airbase.platform.setSystemParameter(SYSPARAM_ROBOT_SPEED, speed);
  airbase.platform.setSystemParameter(SYSPARAM_ROBOT_ANGULAR_SPEED, speed);

  // set the map
  char ch;
  bool base_lock = true;
  airbase.set_baselock_state(base_lock);
  while (true) {
    std::cout << "Please select option:" << std::endl;
    std::cout << "0. unlock the base" << std::endl;
    std::cout << "1. rebuild the map" << std::endl;
    std::cout << "2. load the map from local file(please move car close to the "
                 "origin)"
              << std::endl;
    std::cin >> ch;
    std::cin.ignore();
    if (ch == '0') {
      base_lock = !base_lock;
    } else if (ch == '1') {
      airbase.build_stcm_map(map_path);
      break;
    } else if (ch == '2') {
      airbase.load_stcm_map(map_path);
      break;
    }
    airbase.set_baselock_state(base_lock);
  }

  // begin keyboardControl mode
  initscr();
  noecho();
  cbreak();
  nodelay(stdscr, TRUE);

  bool running = false;

  Direction forward(ACTION_DIRECTION::FORWARD);
  Direction backward(ACTION_DIRECTION::BACKWARD);
  Direction turnleft(ACTION_DIRECTION::TURNLEFT);
  Direction turnright(ACTION_DIRECTION::TURNRIGHT);

  while ((ch = getch()) != 'q') {
    clear();
    printw("Keyboard Control Mode, Press 'q' to quit.\n");
    printw(
        "Press 'w' to move forward\n"
        "Press 's' to move backward\n"
        "Press 'a' to turn left\n"
        "Press 'd' to turn right\n");
    switch (ch) {
      case 'w':
        airbase.platform.moveBy(forward);
        break;
      case 's':
        airbase.platform.moveBy(backward);
        break;
      case 'a':
        airbase.platform.moveBy(turnleft);
        break;
      case 'd':
        airbase.platform.moveBy(turnright);
        break;
      default:
        break;
    }
    boost::this_thread::sleep_for(boost::chrono::milliseconds(2));
  }
  endwin();

  return 0;
}

File: examples/keyboard_control.py
import airbase_py
import sys

params = {
    "SYSPARAM_ROBOT_SPEED": "base.max_moving_speed",
    "SYSPARAM_ROBOT_ANGULAR_SPEED": "base.max_angular_speed",
    "SYSVAL_ROBOT_SPEED_HIGH": "high",
    "SYSVAL_ROBOT_SPEED_MEDIUM": "medium",
    "SYSVAL_ROBOT_SPEED_LOW": "low",
    "SYSPARAM_POWEROFF_WAIT_TIME": "base.power_off_wait_time",
    "SYSPARAM_EMERGENCY_STOP": "base.emergency_stop",
    "SYSVAL_EMERGENCY_STOP_ON": "on",
    "SYSVAL_EMERGENCY_STOP_OFF": "off",
    "SYSPARAM_BRAKE_RELEASE": "base.brake_release",
    "SYSVAL_BRAKE_RELEASE_ON": "on",
    "SYSVAL_BRAKE_RELEASE_OFF": "off",
    "SYSPARAM_DOCKED_REGISTER_STRATEGY": "docking.docked_register_strategy",
    "SYSVAL_DOCKED_REGISTER_STRATEGY_ALWAYS": "always",
    "SYSVAL_DOCKED_REGISTER_STRATEGY_WHEN_NOT_EXISTS": "when_not_exists",
    "SYSPARAM_ACTIONS_WAIT_FOR_ACCEPTABLE_PATH_TIME": "actions.wait_for_acceptable_path_time",
    "SYSPARAM_VIRTUAL_TRACKS_ACCESSIBLE_AREA": "virtual_track.enable_limited_accessible_area",
    "SYSVAL_ACCESSIBLE_AREA_ON": "on",
    "SYSVAL_ACCESSIBLE_AREA_OFF": "off",
}

if __name__ == "__main__":

    sys.argv.append("192.168.11.1")
    airbase = airbase_py.create(sys.argv)
    airbase.platform.set_system_parameter(
        params["SYSPARAM_ROBOT_SPEED"], params["SYSVAL_ROBOT_SPEED_HIGH"]
    )
    airbase.platform.set_system_parameter(
        params["SYSPARAM_ROBOT_ANGULAR_SPEED"], params["SYSVAL_ROBOT_SPEED_HIGH"]
    )

    # set the map
    base_lock = True
    airbase.set_base_lock_state(base_lock)
    while True:
        print("Please select option:")
        print("0. unlock the base")
        print("1. rebuild the map")
        print("2. load the map from local file(please move car close to the origin)")
        key = input("Please input the option and press Enter:")
        if key == "0":
            # 求反
            base_lock = not base_lock
        elif key == "1":
            airbase.build_stcm_map("map.stcm")
            break
        elif key == "2":
            airbase.load_stcm_map("map.stcm")
            break
        airbase.set_base_lock_state(base_lock)

    forward = airbase_py.Direction(airbase_py.ACTION_DIRECTION.FORWARD)
    backward = airbase_py.Direction(airbase_py.ACTION_DIRECTION.BACKWARD)
    turn_left = airbase_py.Direction(airbase_py.ACTION_DIRECTION.TURNLEFT)
    turn_right = airbase_py.Direction(airbase_py.ACTION_DIRECTION.TURNRIGHT)

    while True:
        key = input(
            "Press 'w' 's' 'a' 'd' to move (press Enter to submit), 'q' to quit:"
        )
        if key == "w":
            airbase.move(forward)
        elif key == "s":
            airbase.move(backward)
        elif key == "a":
            airbase.move(turn_left)
        elif key == "d":
            airbase.move(turn_right)
        elif key == "q":
            break
